libdrgn/dwarf_index.c

```diff
@@ -492,22 +516,65 @@ struct drgn_error *drgn_dwarf_index_init(struct drgn_dwarf_index *dindex,
 	return NULL;
 }
 
+static void drgn_dwarf_index_cu_deinit(struct drgn_dwarf_index_cu *cu) {
+	free(cu->file_name_hashes);
+	free(cu->abbrev_insns);
+	free(cu->abbrev_decls);
+}
+
+static void drgn_dwarf_index_deinit_namespaces(struct drgn_dwarf_index_namespace *parent_ns) {
+	const uint64_t ns_tag = DW_TAG_namespace;
+	struct drgn_error *err;
+	struct drgn_dwarf_index_iterator_vector it_vector;
+	struct drgn_dwarf_index_iterator *it;
+	struct drgn_dwarf_index_die *die;
+
+	drgn_dwarf_index_iterator_vector_init(&it_vector);
+
+	it = drgn_dwarf_index_iterator_vector_append_entry(&it_vector);
+	if (!it)
+		// We can't do much on memory failures here.
```

Comment:
Just flagging this - in order to free this completely, we have to allocate some memory (either through recursion or like this). Right now this is a leak because I don't have a better idea (other than making the deallocation something that can fail...). Hopefully this will work out as we just freed a large chunk of the index just before this.

Reply:
I'm on the fence about this vs. recursion. I'm leaning towards recursion so we don't have to do the explicit memory allocation. If we're paranoid, we can limit the depth we're willing to free (or limit it earlier before we create a namespace of that depth), but practically I think we can get by without it.